lib/EnsEMBL/REST/Model/ga4gh/datasets.pm

```diff
@@ -0,0 +1,99 @@
+=head1 LICENSE
+
+Copyright [1999-2014] Wellcome Trust Sanger Institute and the EMBL-European Bioinformatics Institute
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+=cut
+
+package EnsEMBL::REST::Model::ga4gh::datasets;
+
+use Moose;
+extends 'Catalyst::Model';
+
+use Bio::EnsEMBL::IO::Parser::VCF4Tabix;
+use Bio::EnsEMBL::Variation::DBSQL::VCFCollectionAdaptor;
+use EnsEMBL::REST::Model::ga4gh::ga4gh_utils;
+
+use Digest::MD5 qw(md5_hex);
+with 'Catalyst::Component::InstancePerContext';
+
+has 'context' => (is => 'ro');
+use EnsEMBL::REST::Model::ga4gh::ga4gh_utils;
+
+sub build_per_context_instance {
+  my ($self, $c, @args) = @_;
+  return $self->new({ context => $c, %$self, @args });
+}
+
+sub fetch_datasets{
+  my ($self, $data ) = @_;
+
+  my @datasets;
+
+  ## paging
+  my $count = 0;
+  my $nextPageToken;      
+  my $start = 1;
+  $start = 0 if defined $data->{pageToken} && $data->{pageToken} ne ''; 
+
+  # get hash of md5 => description for dataasets
+  my $datasets = $self->context->model('ga4gh::ga4gh_utils')->fetch_all_Datasets();
+
+  foreach my $id( sort keys %{$datasets}){
+
+    $start = 1 if (defined $data->{pageToken} &&  $data->{pageToken} eq $id);
+
+    ## skip datasets returned in last batch
+    next if $start == 0 ;
+
+    ## store start of next batch before exiting this one
+    if ($count == $data->{pageSize}){
+      $nextPageToken = $id;
+      last; 
+    }
+
+     ## save and increment
+     my $dataset = { id => $id, description => $datasets->{$id} };
+     push @datasets, $dataset;
+     $count++;
+  }
+  
+  $self->context()->go( 'ReturnError', 'custom', [ " No datasets are available"])
```

Comment:
Model instructing server flow is bad form - it short-circuits the server. Better to throw exceptions to the controller and handle the errors later.